Validate the report_data field in the attestation test

DomAyre opened the pull request:
Closes https://github.com/microsoft/privacy-sandbox-dev/issues/251

Expands the existing tests which previously only checked that `fetchSnpAttestation()` returned \*something\* for three cases:
- No `report_data` provided (defaults to "")
- `report_data` which is under 64 bytes is provided, providing coverage that we properly zero out the report_data field 
- `report_data` which is over 64 bytes is provided, providing coverage that we properly truncate it.

Now we don't only validate that we get some attestation, we parse out the `report_data` field and confirm it's an exact match with the expected value

Update: we have also made some changes to the implementation being tested to make it more readable

DomAyre committed:
Validate the report_data field in the attestation test

DomAyre commented:
Tested with privacy-sandbox-dev
https://github.com/microsoft/privacy-sandbox-dev/actions/runs/11841619306 ✘

Update: 
- Found issue with data plane precommit
- Fixed in https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/0efa5ce627cc2540150c4c40dace93fd896e93c1

DomAyre committed:
Fix clang format issues

DomAyre commented:
Re run privacy-sandbox-dev tests:
https://github.com/microsoft/privacy-sandbox-dev/actions/runs/11841828614 ✔

KenGordon commented on src/azure/attestation/test/attestation_test.cc:
```diff
@@ -38,24 +81,32 @@ TEST_F(JsonAttestationReportTest, FetchRealAttestation) {
+  std::string report_data = "example_report_data";
+  auto attestation_report = fetchSnpAttestation(toHex(report_data));
+  EXPECT_TRUE(attestation_report.has_value());
+  EXPECT_EQ(getReportData(attestation_report->evidence), report_data);
```
It is not clear to me that this is checking all 64 bytes. Maybe it is but I'd be sure if the line above was something like:
EXPECT_EQ(getReportData(attestation_report->evidence).size(), 64);

DomAyre committed:
Add assert that report_data_str is expected size

DomAyre commented:
Offline discussion with Ken, he was worried the tests might accidentally not consider bytes after the termination character, meaning junk data past that won't be flagged as a problem. 

The test uses the std::string constructor which takes a size so I don't believe it's possible.

To be sure, I've added an assert to ensure it's impossible for bytes to be accidentally ignored when constructing the `report_data` string
https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/0b9b4b79064961eeddd4a6ed8f4ee1f9bafe3563

Accompanying CI run:
https://github.com/microsoft/privacy-sandbox-dev/actions/runs/11843121039

KenGordon reviewed (approved):
Thanks

DomAyre committed:
Remove the use of auto

DomAyre committed:
Add a comment explaining SnpRequest initialisation

DomAyre committed:
Add a comment about calculating the bytes to copy

DomAyre committed:
Compare the output with right padded input

DomAyre committed:
Remove use of auto in attestation test

DomAyre committed:
Use hard coded byte arrays for test checks

DomAyre committed:
Add tests for getReport

DomAyre commented:
There was further offline discussion and as a result the following changes have been requested:

In `sev6.cc`
- Swap auto for explicit types https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/ec573db10821350c532b7efa847fcef3fc6707ed
- Add a comment explaining the purpose of the calculation for the size used in the copy https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/497c382265cc8a038ae3f77fdaee0d7762943c89
- Add a comment explaining the initialiser of SnpRequest https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/0975c14b4db1eff9d6702b29d76354e6484fec0b

In `attestation_test.cc`
- Swap auto for explicit types https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/b85ef756170e2bf8a4e1b35d05594e99ec0e6070
- Add tests to call sev6.cc::getReport() as well as attestation.h::fetchSnpAttestation() https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/55e0e9db9f1155fb072403a38cd8c07f73c73a09
- Hard code the expected bytes in the test as a separate variable for comparison https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/2295284e25863adcefb7e537b62c8ed8c08d2365

Accompanying CI run: https://github.com/microsoft/privacy-sandbox-dev/actions/runs/11855446577

KenGordon commented on src/azure/attestation/test/attestation_test.cc:
```diff
@@ -134,4 +136,64 @@ TEST_F(JsonAttestationReportTest, FetchRealAttestationNonSnp) {
+  }
+  // Define test inputs and outputs
+  std::string report_data = "";
+  uint8_t expected[64] = {
```
std::array<uint8_t, 64>

DomAyre replied:
Yes for these tests since the ask was about minimal transformations on the inputs or outputs I kept as uint8_t[64] but I'm very pro using std::array so I'll make that change

DomAyre replied:
Addressed in https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/c747c834323474ad2c624c92470f40f59f729099

KenGordon commented on src/azure/attestation/test/attestation_test.cc:
```diff
@@ -134,4 +136,64 @@ TEST_F(JsonAttestationReportTest, FetchRealAttestationNonSnp) {
+  };
+
+  std::unique_ptr<SnpReport> snp_report = getReport(report_data);
+
```
another std::array constructed from snp_report.report_data

DomAyre replied:
Addressed in https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/c747c834323474ad2c624c92470f40f59f729099

KenGordon commented on src/azure/attestation/test/attestation_test.cc:
```diff
@@ -134,4 +136,64 @@ TEST_F(JsonAttestationReportTest, FetchRealAttestationNonSnp) {
+
+  std::unique_ptr<SnpReport> snp_report = getReport(report_data);
+
+  EXPECT_TRUE(std::equal(std::begin(expected), std::end(expected),
```
std::array operator == should manage this such that EXPECT_EQ(expected_report, resulting_report) is good, conside and make sense quickly. It also avoids having to thing about the scope of the result of the get() on the unique ptr.

KenGordon replied:
If fact you seem to have done that at the top

DomAyre replied:
Yes I am pro using std::array I only chose uint8_t here because I thought you wanted minimal transformations or helper code so I'm happy to make that change

DomAyre replied:
Addressed in https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/c747c834323474ad2c624c92470f40f59f729099

DomAyre committed:
Add a test with report_data with midway null character

KenGordon commented:
A std::array over a uint8_t[64] is not really a transformation. The array ctor is just a byte for byte copy and in the case of the expected[64] if you get the ctor correct the compiler might make it all static anyway (although I'd not try too hard with constexpr etc as we really do not care about code size or exec time here.


________________________________
From: Dominic Ayre ***@***.***>
Sent: 15 November 2024 13:14
To: KenGordon/data-plane-shared-libraries ***@***.***>
Cc: Ken Gordon ***@***.***>; Review requested ***@***.***>
Subject: Re: [KenGordon/data-plane-shared-libraries] Validate the report_data field in the attestation test (PR #49)


@DomAyre commented on this pull request.

________________________________

In src/azure/attestation/test/attestation_test.cc<https://github.com/KenGordon/data-plane-shared-libraries/pull/49#discussion_r1843738813>:

> +
+TEST_F(JsonAttestationReportTest, GetReport) {
+  if (!hasSnp()) {
+    return;
+  }
+  // Define test inputs and outputs
+  std::string report_data = "";
+  uint8_t expected[64] = {
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  };
+
+  std::unique_ptr<SnpReport> snp_report = getReport(report_data);
+
+  EXPECT_TRUE(std::equal(std::begin(expected), std::end(expected),


Yes I am pro using std::array I only chose uint8_t here because I thought you wanted minimal transformations or helper code so I'm happy to make that change

—
Reply to this email directly, view it on GitHub<https://github.com/KenGordon/data-plane-shared-libraries/pull/49#discussion_r1843738813>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABI7RIABPA4OWHPOB3YYK7L2AXXSDAVCNFSM6AAAAABRZKHXW6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDIMZYGUZTCMJXGQ>.
You are receiving this because your review was requested.Message ID: ***@***.***>

DomAyre committed:
Use std::array in all tests

DomAyre commented:
Added a couple of updates:

- Added a test which uses a prebaked hex string with a null character in the middle https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/e8b240ffb6a26eba318f7dfeca776d519e281522
- Updated the tests for `getReport()` to also use std::array https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/c747c834323474ad2c624c92470f40f59f729099

Accompanying CI run: https://github.com/microsoft/privacy-sandbox-dev/actions/runs/11857543834

DomAyre committed:
Add a test with a non-char byte

DomAyre commented:
CI run for https://github.com/KenGordon/data-plane-shared-libraries/pull/49/commits/37db1587e66cc87fb9f9e4c3cd0b9c71532cd6b2:
https://github.com/microsoft/privacy-sandbox-dev/actions/runs/11858021420
